flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/FixedDelayRestartBackoffTimeStrategy.java

```diff
@@ -61,8 +61,9 @@ public long getBackoffTime() {
     }
 
     @Override
-    public void notifyFailure(Throwable cause) {
+    public boolean notifyFailure(Throwable cause) {
         currentRestartAttempt++;
+        return true;
```

Comment:
I understand that it wasn't covered by FLIP-364. But from a conceptual point of view: shouldn't we cover it also for the two other restart strategies. Or am I missing something here? :thinking:  I'm also find with having this being covered by follow-up Jira issues.

Reply:
> But from a conceptual point of view: shouldn't we cover it also for the two other restart strategies. Or am I missing something here?

In the beginning, I want to improve all restart strategies, but I meet some feedback to removing other restart strategies during discussion. The core background is this thread: https://lists.apache.org/thread/l7wyc7pndpsvh2h7hj3fw2td9yphrlox

In brief, 3 reasons:

1. The semantics of option

    - the failure-rate strategy's restart upper limit option is named  `restart-strategy.failure-rate.max-failures-per-interval`
    - It's  `max-failures-per-interval` instead of `max-attempts-per-interval`.
    - If we improve it directly, the name and behaviour aren't matched.
2. We recommend users use the `exponential-delay restart strategy` in the future, it's more powerful.
3. After FLIP-360 discussion, I found `exponential-delay restart strategy` can replace other restart strategies directly if users set the `restart-strategy.exponential-delay.backoff-multiplier` = 1

Actually, I think other restart-strategies can be deprecated in the future.

Reply:
Thanks for the clarification. That makes sense.

> After FLIP-360 discussion, I found exponential-delay restart strategy can replace other restart strategies directly if users set the restart-strategy.exponential-delay.backoff-multiplier = 1

I guess, you meant [FLIP-364](https://cwiki.apache.org/confluence/display/FLINK/FLIP-364%3A+Improve+the+exponential-delay+restart-strategy) here? ;-) You're right, we could deprecate the class. But there is still a benefit of keeping the configuration parameter (as an alias that configures the `ExponentialDelayRestartBackoffTimeStrategy` with multiplier=1) to make it easier to configure this behavior. 

What about deprecating that class as part of this PR and coming up with a follow-up Jira issue that replaces the strategy?

Reply:
> I guess, you meant [FLIP-364](https://cwiki.apache.org/confluence/display/FLINK/FLIP-364%3A+Improve+the+exponential-delay+restart-strategy) here?

Exactly, sorry for my typo.

> What about deprecating that class as part of this PR and coming up with a follow-up Jira issue that replaces the strategy?

They are internal class, IIUC, we can refactor it directly without the `@Deprecated` annotation.

---------------------------------------------------------------------------------

Your suggestion is that we still keep the failure-rate and fixed-delay restart-strategies, but we can reuse the `ExponentialDelayRestartBackoffTimeStrategy` class, right?

I'm thinking could we deprecate failure-rate and fixed-delay restart-strategies directly? Users can configure them directly.

Keep failure-rate and fixed-delay restart-strategies and reusing `ExponentialDelayRestartBackoffTimeStrategy` class still has the semantic problem that I mentioned in the last comment [1].

[1] https://github.com/apache/flink/pull/24003#discussion_r1456807889